### PR TITLE
Go modules

### LINKED ISSUE #41: golang: use go modules
We'll eventually need to upgrade dependency management to [go modules](https://github.com/golang/go/wiki/Modules). Maybe this will happen upstream reasonably soon, or we'll have to move ahead independently. I think this feature will get turned `on` by default with go1.13, which is scheduled I believe for August 2019. 

- https://blog.golang.org/modules2019
- https://github.com/golang/go/wiki/Modules

### PR DESCRIPTION
**Pull request by tzdybal**
This PR replaces `env.sh` script with Go Modules. It's mostly backward compatible:
 - you can build in `$GOPATH/src/github.com/ethereum/go-ethereum`
 - you can build in any other directory outside `$GOPATH`

The one "incompatibility" is:
 - to build in `$GOPATH/src/github.com/multi-geth/multi-geth` you need to set environment variable `GO111MODULES=on` - this will be default behavior in Go 1.13.

This PR resolves #41.

Hopefully this changes will be accepted as is in upstream repository.

### CODE REVIEW

**Review by r8d8 — APPROVED** [review]

**Comment by r8d8** [comment]
```
go run build/ci.go lint
>>> /home/travis/.gimme/versions/go1.11.10.linux.amd64/bin/go get gopkg.in/alecthomas/kingpin.v3-unstable@63abe20a23e29e80bbef8089bd3dee3ac25e5306
go get: warning: modules disabled by GO111MODULE=auto in GOPATH/src;
	ignoring go.mod;
	see 'go help modules'
```
need to adjust travis lint check

**Comment by meowsbits** [comment]
@tzdybal Did you PR this upstream? If not, :pray:

**Comment by tzdybal** [comment]
Fixed problem with `make lint`. KISS. Please re-review.

@meowsbits I'll make upstream PR after merging this PR.

**Comment by meowsbits** [comment]
I did a fresh (new, random location) clone to try this 'from scratch'. When I ran `make all` I saw an error, but was subsequently unable to reproduce.

I don't think it's a big deal, since I haven't been able to reproduce. Maybe it was an internet or otherwise source availability issue?

```
$ make all                                                                                                                                                
go run build/ci.go install                                                                                                                                
>>> /home/ia/go1.12.1.linux-amd64/go/bin/go install -ldflags -X main.gitCommit=a5535cc409bef389dffda49a2d051ed44956f0be -v go: downloading bazil.org/fuse 
v0.0.0-20160811212531-371fbbdaa898 go: downloading github.com/olekukonko/tablewriter v0.0.0-20170128050532-febf2d34b54a go: downloading github.com/uber/ja
eger-client-go v0.0.0-20180607151842-f7e0d4744fa6 go: downloading github.com/julienschmidt/httprouter v0.0.0-20170430222011-975b5c4c7c21 go: downloading g
ithub.com/docker/docker v0.0.0-20180625184442-8e610b2b55bf go: extracting github.com/olekukonko/tablewriter v0.0.0-20170128050532-febf2d34b54a go: extract
ing github.com/julienschmidt/httprouter v0.0.0-20170430222011-975b5c4c7c21 go: extracting github.com/uber/jaeger-client-go v0.0.0-20180607151842-f7e0d4744
fa6 go: downloading github.com/uber/jaeger-lib v0.0.0-20180615202729-a51202d6f4a7 go: extracting github.com/uber/jaeger-lib v0.0.0-20180615202729-a51202d6
f4a7 go: downloading github.com/codahale/hdrhistogram v0.0.0-20161010025455-3a0bb77429bd go: extracting github.com/codahale/hdrhistogram v0.0.0-2016101002
5455-3a0bb77429bd go: extracting bazil.org/fuse v0.0.0-20160811212531-371fbbdaa898 go: extracting github.com/docker/docker v0.0.0-20180625184442-8e610b2b5
5bf github.com/ethereum/go-ethereum github.com/ethereum/go-ethereum/accounts github.com/ethereum/go-ethereum/accounts/abi github.com/ethereum/go-ethereum/
accounts/abi/bind github.com/ethereum/go-ethereum/accounts/abi/bind/backends github.com/ethereum/go-ethereum/accounts/external github.com/ethereum/go-ethe
reum/accounts/keystore github.com/ethereum/go-ethereum/accounts/scwallet github.com/ethereum/go-ethereum/accounts/usbwallet github.com/ethereum/go-ethereu
m/accounts/usbwallet/trezor github.com/ethereum/go-ethereum/cmd/abigen github.com/ethereum/go-ethereum/cmd/bootnode github.com/ethereum/go-ethereum/cmd/cl
ef github.com/ethereum/go-ethereum/cmd/ethkey github.com/ethereum/go-ethereum/cmd/evm github.com/ethereum/go-ethereum/cmd/evm/internal/compiler github.com
/ethereum/go-ethereum/cmd/faucet github.com/ethereum/go-ethereum/cmd/geth github.com/ethereum/go-ethereum/cmd/p2psim github.com/ethereum/go-ethereum/cmd/p
uppeth github.com/ethereum/go-ethereum/cmd/rlpdump github.com/ethereum/go-ethereum/cmd/swarm github.com/ethereum/go-ethereum/cmd/swarm/global-store github
.com/ethereum/go-ethereum/cmd/swarm/mimegen github.com/ethereum/go-ethereum/cmd/swarm/swarm-smoke github.com/ethereum/go-ethereum/cmd/swarm/swarm-snapshot
 github.com/ethereum/go-ethereum/cmd/utils github.com/ethereum/go-ethereum/cmd/wnode github.com/ethereum/go-ethereum/common github.com/ethereum/go-ethereu
m/common/bitutil github.com/ethereum/go-ethereum/common/compiler github.com/ethereum/go-ethereum/common/fdlimit github.com/ethereum/go-ethereum/common/hex
util github.com/ethereum/go-ethereum/common/math github.com/ethereum/go-ethereum/common/mclock github.com/ethereum/go-ethereum/common/prque github.com/eth
ereum/go-ethereum/consensus github.com/ethereum/go-ethereum/consensus/clique github.com/ethereum/go-ethereum/consensus/ethash github.com/ethereum/go-ether
eum/consensus/misc github.com/ethereum/go-ethereum/console github.com/ethereum/go-ethereum/contracts/chequebook github.com/ethereum/go-ethereum/contracts/
chequebook/contract github.com/ethereum/go-ethereum/contracts/ens github.com/ethereum/go-ethereum/contracts/ens/contract github.com/ethereum/go-ethereum/c
ontracts/ens/fallback_contract github.com/ethereum/go-ethereum/core github.com/ethereum/go-ethereum/core/asm github.com/ethereum/go-ethereum/core/bloombit
s github.com/ethereum/go-ethereum/core/rawdb github.com/ethereum/go-ethereum/core/state github.com/ethereum/go-ethereum/core/types github.com/ethereum/go-
ethereum/core/vm github.com/ethereum/go-ethereum/core/vm/runtime github.com/ethereum/go-ethereum/crypto github.com/ethereum/go-ethereum/crypto/bn256 githu
b.com/ethereum/go-ethereum/crypto/bn256/cloudflare github.com/ethereum/go-ethereum/crypto/bn256/google github.com/ethereum/go-ethereum/crypto/ecies github
.com/ethereum/go-ethereum/crypto/secp256k1 github.com/ethereum/go-ethereum/dashboard github.com/ethereum/go-ethereum/eth github.com/ethereum/go-ethereum/e
th/downloader github.com/ethereum/go-ethereum/eth/fetcher github.com/ethereum/go-ethereum/eth/filters github.com/ethereum/go-ethereum/eth/gasprice github.
com/ethereum/go-ethereum/eth/tracers github.com/ethereum/go-ethereum/eth/tracers/internal/tracers github.com/ethereum/go-ethereum/ethclient github.com/eth
ereum/go-ethereum/ethdb github.com/ethereum/go-ethereum/ethdb/leveldb github.com/ethereum/go-ethereum/ethdb/memorydb github.com/ethereum/go-ethereum/ethst
ats github.com/ethereum/go-ethereum/event github.com/ethereum/go-ethereum/graphql github.com/ethereum/go-ethereum/internal/build github.com/ethereum/go-et
hereum/internal/cmdtest github.com/ethereum/go-ethereum/internal/debug github.com/ethereum/go-ethereum/internal/ethapi github.com/ethereum/go-ethereum/int
ernal/guide github.com/ethereum/go-ethereum/internal/jsre github.com/ethereum/go-ethereum/internal/jsre/deps github.com/ethereum/go-ethereum/internal/open
rpc github.com/ethereum/go-ethereum/internal/web3ext github.com/ethereum/go-ethereum/les github.com/ethereum/go-ethereum/les/flowcontrol github.com/ethere
um/go-ethereum/light github.com/ethereum/go-ethereum/log github.com/ethereum/go-ethereum/metrics github.com/ethereum/go-ethereum/metrics/exp github.com/et
hereum/go-ethereum/metrics/influxdb github.com/ethereum/go-ethereum/metrics/librato github.com/ethereum/go-ethereum/metrics/prometheus github.com/ethereum
/go-ethereum/miner github.com/ethereum/go-ethereum/mobile github.com/ethereum/go-ethereum/node github.com/ethereum/go-ethereum/p2p github.com/ethereum/go-
ethereum/p2p/discover github.com/ethereum/go-ethereum/p2p/discv5 github.com/ethereum/go-ethereum/p2p/enode github.com/ethereum/go-ethereum/p2p/enr github.
com/ethereum/go-ethereum/p2p/nat github.com/ethereum/go-ethereum/p2p/netutil github.com/ethereum/go-ethereum/p2p/protocols github.com/ethereum/go-ethereum
/p2p/simulations github.com/ethereum/go-ethereum/p2p/simulations/adapters github.com/ethereum/go-ethereum/p2p/simulations/examples github.com/ethereum/go-
ethereum/p2p/simulations/pipes github.com/ethereum/go-ethereum/p2p/testing github.com/ethereum/go-ethereum/params github.com/ethereum/go-ethereum/rlp gith
ub.com/ethereum/go-ethereum/rpc github.com/ethereum/go-ethereum/signer/core github.com/ethereum/go-ethereum/signer/fourbyte github.com/ethereum/go-ethereu
m/signer/rules github.com/ethereum/go-ethereum/signer/rules/deps github.com/ethereum/go-ethereum/signer/storage github.com/ethereum/go-ethereum/swarm gith
ub.com/ethereum/go-ethereum/swarm/api github.com/ethereum/go-ethereum/swarm/api/client github.com/ethereum/go-ethereum/swarm/api/http github.com/ethereum/
go-ethereum/swarm/bmt github.com/ethereum/go-ethereum/swarm/chunk github.com/ethereum/go-ethereum/swarm/fuse github.com/ethereum/go-ethereum/swarm/log git
hub.com/ethereum/go-ethereum/swarm/metrics github.com/ethereum/go-ethereum/swarm/network github.com/ethereum/go-ethereum/swarm/network/bitvector github.co
m/ethereum/go-ethereum/swarm/network/priorityqueue github.com/ethereum/go-ethereum/swarm/network/simulation github.com/ethereum/go-ethereum/swarm/network/
simulations github.com/ethereum/go-ethereum/swarm/network/simulations/discovery github.com/ethereum/go-ethereum/swarm/network/stream github.com/ethereum/g
o-ethereum/swarm/network/stream/intervals github.com/ethereum/go-ethereum/swarm/pot github.com/ethereum/go-ethereum/swarm/pss github.com/ethereum/go-ether
eum/swarm/pss/client github.com/ethereum/go-ethereum/swarm/pss/notify github.com/ethereum/go-ethereum/swarm/sctx github.com/ethereum/go-ethereum/swarm/ser
vices/swap github.com/ethereum/go-ethereum/swarm/services/swap/swap github.com/ethereum/go-ethereum/swarm/shed github.com/ethereum/go-ethereum/swarm/spanc
ontext github.com/ethereum/go-ethereum/swarm/state github.com/ethereum/go-ethereum/swarm/storage github.com/ethereum/go-ethereum/swarm/storage/encryption 
github.com/ethereum/go-ethereum/swarm/storage/feed github.com/ethereum/go-ethereum/swarm/storage/feed/lookup github.com/ethereum/go-ethereum/swarm/storage
/localstore github.com/ethereum/go-ethereum/swarm/storage/mock github.com/ethereum/go-ethereum/swarm/storage/mock/db github.com/ethereum/go-ethereum/swarm
/storage/mock/explorer github.com/ethereum/go-ethereum/swarm/storage/mock/mem github.com/ethereum/go-ethereum/swarm/storage/mock/rpc github.com/ethereum/g
o-ethereum/swarm/storage/mock/test github.com/ethereum/go-ethereum/swarm/swap github.com/ethereum/go-ethereum/swarm/testutil github.com/ethereum/go-ethere
um/swarm/tracing github.com/ethereum/go-ethereum/swarm/version github.com/ethereum/go-ethereum/tests github.com/ethereum/go-ethereum/trie github.com/ether
eum/go-ethereum/whisper/mailserver github.com/ethereum/go-ethereum/whisper/shhclient github.com/ethereum/go-ethereum/whisper/whisperv6                    
https fetch failed: parse https://go: extracting github.com/codahale/hdrhistogram v0.0.0-20161010025455-3a0bb77429bd: invalid character " " in host name  
https fetch failed: parse https://go: extracting github.com/codahale: invalid character " " in host name                                                  
https fetch failed: parse https://go: extracting github.com: invalid character " " in host name                                                           
https fetch failed: parse https://go: downloading github.com/codahale/hdrhistogram v0.0.0-20161010025455-3a0bb77429bd: invalid character " " in host name 
https fetch failed: parse https://go: downloading github.com/codahale: invalid character " " in host name                                                 
https fetch failed: parse https://go: downloading github.com: invalid character " " in host name                                                          
https fetch failed: parse https://go: downloading bazil.org/fuse v0.0.0-20160811212531-371fbbdaa898: invalid character " " in host name
https fetch failed: parse https://go: downloading bazil.org: invalid character " " in host name                                                           
https fetch failed: parse https://go: downloading github.com/julienschmidt/httprouter v0.0.0-20170430222011-975b5c4c7c21: invalid character " " in host n$
me   
https fetch failed: parse https://go: downloading github.com/julienschmidt: invalid character " " in host name
https fetch failed: parse https://go: extracting bazil.org/fuse v0.0.0-20160811212531-371fbbdaa898: invalid character " " in host name
https fetch failed: parse https://go: extracting bazil.org: invalid character " " in host name
https fetch failed: parse https://go: downloading github.com/uber/jaeger-client-go v0.0.0-20180607151842-f7e0d4744fa6: invalid character " " in host name
https fetch failed: parse https://go: downloading github.com/uber: invalid character " " in host name
https fetch failed: parse https://go: downloading github.com/uber/jaeger-lib v0.0.0-20180615202729-a51202d6f4a7: invalid character " " in host name
https fetch failed: parse https://go: extracting github.com/julienschmidt/httprouter v0.0.0-20170430222011-975b5c4c7c21: invalid character " " in host na$
e                                                                                                                                    
https fetch failed: parse https://go: extracting github.com/julienschmidt: invalid character " " in host name
https fetch failed: parse https://go: downloading github.com/docker/docker v0.0.0-20180625184442-8e610b2b55bf: invalid character " " in host name
https fetch failed: parse https://go: downloading github.com/docker: invalid character " " in host name
https fetch failed: parse https://go: extracting github.com/docker/docker v0.0.0-20180625184442-8e610b2b55bf: invalid character " " in host name
https fetch failed: parse https://go: extracting github.com/docker: invalid character " " in host name
https fetch failed: parse https://go: extracting github.com/olekukonko/tablewriter v0.0.0-20170128050532-febf2d34b54a: invalid character " " in host name 
https fetch failed: parse https://go: extracting github.com/olekukonko: invalid character " " in host name                                                
https fetch failed: parse https://go: extracting github.com/uber/jaeger-lib v0.0.0-20180615202729-a51202d6f4a7: invalid character " " in host name        
https fetch failed: parse https://go: extracting github.com/uber: invalid character " " in host name
https fetch failed: parse https://go: downloading github.com/olekukonko/tablewriter v0.0.0-20170128050532-febf2d34b54a: invalid character " " in host name
https fetch failed: parse https://go: downloading github.com/olekukonko: invalid character " " in host name
https fetch failed: parse https://go: extracting github.com/uber/jaeger-client-go v0.0.0-20180607151842-f7e0d4744fa6: invalid character " " in host name
can't load package: package go: downloading bazil.org/fuse v0.0.0-20160811212531-371fbbdaa898: unknown import path "go: downloading bazil.org/fuse v0.0.0-
20160811212531-371fbbdaa898": cannot find module providing package go: downloading bazil.org/fuse v0.0.0-20160811212531-371fbbdaa898
can't load package: package go: downloading github.com/olekukonko/tablewriter v0.0.0-20170128050532-febf2d34b54a: unknown import path "go: downloading git
hub.com/olekukonko/tablewriter v0.0.0-20170128050532-febf2d34b54a": cannot find module providing package go: downloading github.com/olekukonko/tablewriter
 v0.0.0-20170128050532-febf2d34b54a
can't load package: package go: downloading github.com/uber/jaeger-client-go v0.0.0-20180607151842-f7e0d4744fa6: unknown import path "go: downloading gith
ub.com/uber/jaeger-client-go v0.0.0-20180607151842-f7e0d4744fa6": cannot find module providing package go: downloading github.com/uber/jaeger-client-go v0
.0.0-20180607151842-f7e0d4744fa6
can't load package: package go: downloading github.com/julienschmidt/httprouter v0.0.0-20170430222011-975b5c4c7c21: unknown import path "go: downloading g
ithub.com/julienschmidt/httprouter v0.0.0-20170430222011-975b5c4c7c21": cannot find module providing package go: downloading github.com/julienschmidt/http
router v0.0.0-20170430222011-975b5c4c7c21
can't load package: package go: downloading github.com/docker/docker v0.0.0-20180625184442-8e610b2b55bf: unknown import path "go: downloading github.com/d
ocker/docker v0.0.0-20180625184442-8e610b2b55bf": cannot find module providing package go: downloading github.com/docker/docker v0.0.0-20180625184442-8e61
0b2b55bf
can't load package: package go: extracting github.com/olekukonko/tablewriter v0.0.0-20170128050532-febf2d34b54a: unknown import path "go: extracting githu
b.com/olekukonko/tablewriter v0.0.0-20170128050532-febf2d34b54a": cannot find module providing package go: extracting github.com/olekukonko/tablewriter v0
.0.0-20170128050532-febf2d34b54a
can't load package: package go: extracting github.com/julienschmidt/httprouter v0.0.0-20170430222011-975b5c4c7c21: unknown import path "go: extracting git
hub.com/julienschmidt/httprouter v0.0.0-20170430222011-975b5c4c7c21": cannot find module providing package go: extracting github.com/julienschmidt/httprou
ter v0.0.0-20170430222011-975b5c4c7c21
can't load package: package go: extracting github.com/uber/jaeger-client-go v0.0.0-20180607151842-f7e0d4744fa6: unknown import path "go: extracting github
.com/uber/jaeger-client-go v0.0.0-20180607151842-f7e0d4744fa6": cannot find module providing package go: extracting github.com/uber/jaeger-client-go v0.0.
0-20180607151842-f7e0d4744fa6
can't load package: package go: downloading github.com/uber/jaeger-lib v0.0.0-20180615202729-a51202d6f4a7: unknown import path "go: downloading github.com
/uber/jaeger-lib v0.0.0-20180615202729-a51202d6f4a7": cannot find module providing package go: downloading github.com/uber/jaeger-lib v0.0.0-2018061520272
9-a51202d6f4a7
can't load package: package go: extracting github.com/uber/jaeger-lib v0.0.0-20180615202729-a51202d6f4a7: unknown import path "go: extracting github.com/u
ber/jaeger-lib v0.0.0-20180615202729-a51202d6f4a7": cannot find module providing package go: extracting github.com/uber/jaeger-lib v0.0.0-20180615202729-a
51202d6f4a7
```

File: build/ci.go [hunk 1]
```diff
@@ -357,8 +357,10 @@ func doLint(cmdline []string) {
 		packages = flag.CommandLine.Args()
 	}
 	// Get metalinter and install all supported linters
-	build.MustRun(goTool("get", "gopkg.in/alecthomas/gometalinter.v2"))
-	build.MustRunCommand(filepath.Join(GOBIN, "gometalinter.v2"), "--install")
+	os.Setenv("GO111MODULE", "on")
+	build.MustRun(goTool("get", "gopkg.in/alecthomas/kingpin.v3-unstable@63abe20a23e29e80bbef8089bd3dee3ac25e5306"))
```
Comment (meowsbits) [inline build/ci.go]: What's the reason for this version bump?

Comment (tzdybal) [reply build/ci.go]: I like to use up to date software.

Comment (meowsbits) [reply build/ci.go]: lol. weird.

**Comment by meowsbits** [comment]
```
$ make lint
go run build/ci.go lint
>>> /home/ia/go1.12.1.linux-amd64/go/bin/go get gopkg.in/alecthomas/gometalinter.v2
>>> /home/ia/sandbox/multi-geth/build/bin/gometalinter.v2 --install
Installing:
  deadcode
  dupl
  errcheck
  gochecknoglobals
  gochecknoinits
  goconst
  gocyclo
  goimports
  golint
  gosec
  gosimple
  gotype
  gotypex
  ineffassign
  interfacer
  lll
  maligned
  megacheck
  misspell
  nakedret
  safesql
  staticcheck
  structcheck
  unconvert
  unparam
  unused
  varcheck
can't load package: package github.com/securego/gosec/cmd/gosec: cannot find package "github.com/securego/gosec/cmd/gosec" in any of:
        /home/ia/go1.12.1.linux-amd64/go/src/github.com/securego/gosec/cmd/gosec (from $GOROOT)
        /home/ia/go/src/github.com/alecthomas/gometalinter/_linters/src/github.com/securego/gosec/cmd/gosec (from $GOPATH)
WARNING: failed to install one or more linters: exit status 1 (installing individually)
can't load package: package github.com/securego/gosec/cmd/gosec: cannot find package "github.com/securego/gosec/cmd/gosec" in any of:
        /home/ia/go1.12.1.linux-amd64/go/src/github.com/securego/gosec/cmd/gosec (from $GOROOT)
        /home/ia/go/src/github.com/alecthomas/gometalinter/_linters/src/github.com/securego/gosec/cmd/gosec (from $GOPATH)
WARNING: failed to install github.com/securego/gosec/cmd/gosec: exit status 1
gometalinter.v2: error: failed to install the following linters: github.com/securego/gosec/cmd/gosec
util.go:45: exit status 1
exit status 1
Makefile:41: recipe for target 'lint' failed
make: *** [lint] Error 1
```

**Review by meowsbits — APPROVED** [review]

**Comment by tzdybal** [comment]
Upstream PR is ready for go modules change.

**Pull request closed without merging**